source/extensions/tracers/xray/tracer.cc

```diff
@@ -1,37 +1,134 @@
 #include "extensions/tracers/xray/tracer.h"
 
+#include <sys/socket.h>
+
 #include <algorithm>
 #include <array>
+#include <chrono>
 #include <string>
 
 #include "envoy/http/header_map.h"
+#include "envoy/network/listener.h"
 
+#include "common/common/assert.h"
 #include "common/common/fmt.h"
 #include "common/common/hex.h"
+#include "common/runtime/runtime_impl.h"
+
+#include "source/extensions/tracers/xray/daemon.pb.h"
 
 namespace Envoy {
 namespace Extensions {
 namespace Tracers {
 namespace XRay {
 
+namespace {
+constexpr auto XRaySerializationVersion = "1";
+
+// X-Ray Trace ID Format
+//
+// A trace_id consists of three parts separated by hyphens.
+// For example, 1-58406cf0-a006649127e371903a2de979.
+// This includes:
+//
+// - The version number, that is, 1.
+// - The time of the original request, in Unix epoch time, in 8 hexadecimal digits.
+// - A 96-bit unique identifier in 24 hexadecimal digits.
+//
+// For more details see:
+// https://docs.aws.amazon.com/xray/latest/devguide/xray-api-segmentdocuments.html#api-segmentdocuments-fields
+std::string generateTraceId(Envoy::SystemTime point_in_time) {
```

Comment:
Isn't it more conventional to put this in a utility class?

Reply:
Also I don't think this function has test coverage. Shouldn't we inject the RNG?

Reply:
>Isn't it more conventional to put this in a utility class?

But is it useful to put it in a utility class?
it's used only for x-ray trace IDs (happens only in this file).
Also, because it's not in a class, I'm able to mark it as "internal linkage".

> Also I don't think this function has test coverage. 

It _does_ have coverage by means of testing the functions calling it.

>Shouldn't we inject the RNG?

I'm not sure what the downsides are of not injecting it.

Reply:
I guess I didn't see a test that explicitly tests the output to validate that it returns the 24 character hyphen-delimited output in the correct format. Im more worried about making sure there it a regression test specifically for that.

Reply:
I've added a test.